# Assessment 1:
## Get the square root of 2 and display to 100 decimal places


### Option 1: Newton Raphson method

In [6]:
import decimal as Dec
import datetime as dt

Time_1 = dt.datetime.now()
# Dec.getcontext().prec = 100

Number = Dec.Decimal(2)
bStart = 1                                                                  # Start value is 1
zStart = 1                                                                  # Start value is 1 (x0)
cFinal = 0                                                                  # Variable to continue/end while loop

aCount = 0                                                                  # Iteration count
yNewNum = 0                                                                 # New value of estimate (x1)
xRoot = 0                                                                   # Used to transfer values between yNewNum and zStart

while cFinal != 1:                                                          # While the difference between zStart and yNewNum is greater than 0.0000001 (7 decimal places)
    yNewNum = (zStart + (Number / zStart)) / 2                              # x1 = (x0 + (S/x0)) / 2 
    zStart = yNewNum                                                        # Replace the value x0 with the value of x1
    yNewNum = xRoot                                                         # Replace the value of x1 with the transfer value

    if abs(((zStart - xRoot) / zStart)) < (1/(100**100)):                   # When there is no significant change in the 100^100 decimal place
        cFinal = 1                                                          # Switch off continuance veriable
          
    xRoot = zStart                                                          # Replace the transfer value with x0

    aCount += 1                                                             # aCount iterates by 1

print("Square root 2 to 100 decimal places (Newton-Raphson method): \n", Dec.Decimal(zStart))

Time_2 = dt.datetime.now()
print("Time taken to run:", (Time_2-Time_1))


Square root 2 to 100 decimal places (Newton-Raphson method): 
 1.414213562373095048801688724
Time taken to run: 0:00:00


### Option 2: Alternative method (function)
Soure: https://rosettacode.org/wiki/Integer_roots#Python

In [2]:
def root(a, b):
    if b < 2:
        return b
    a1 = a - 1
    c = 1
    d = (a1 * c + b // (c ** a1)) // a
    e = (a1 * d + b // (d ** a1)) // a
    while c not in (d, e):
        c, d, e = d, e, (a1 * e + b // (e ** a1)) // a
    return min(d, e)

result = str(root(2, 2 * 100 ** 100))
formatted_result = f'{result[0]}.{result[1:]}'
print("Square root 2 to 100 decimal places:")
print(formatted_result)
                                                                 

Square root 2 to 100 decimal places:
1.4142135623730950488016887242096980785696718753769480731766797379907324784621070388503875343276415727


### Option 3: Subtraction method
Source: http://www.afjarvis.staff.shef.ac.uk/maths/jarvisspec02.pdf
https://www.mathblog.dk/project-euler-80-digits-irrational-square-roots/

Square Roots by Subtraction
FRAZER JARVIS
pages 119 - 122
Mathematical Spectrum 2004/2005 Volume 37 Number 3

Japanese approach. Slower than Newton-Raphson.
Newton-Raphson takes 8 steps to run, in near instaneous time.
Subtraction method takes 576 iterations, and up to 500ths of a second to run

In [19]:
def sub_root(a, b):                                     # Create a function to import values of a and b.
    while len(str(b)) <= 102:                           # While the string length of b is less than 101 (1 integter, plus the 101 significant decimal place).

        if a >= b:                                      # If input "a" is greater than or equal input "b".
            a = a - b                                   # Set value of "a" to be "a" minus "b".
            b = b + 10                                  # Set value of "b" to be "b" + 10.

        elif a < b:                                     # If input "a" is less than input "b".
            a = a * 100                                 # Multiply "a" by 100.
            c = len(str(b))-1                           # Set the value "c" to be the string length of "b" minus 1.
            d = str(b)                                  # Set the value "d" to be the string "b".
            e = int(d[:c] + "0" + d[c:])                # Set the value of "e" to be the first "c" number of characters of the string "d".
                                                        # concatenated with a "0" and finished with the last digit of the string "d".
            b = e                                       # Set the value of "b" to be the same as "e".

        return sub_root(a, b)                           # Return the value to the function to reprocess.

    return b                                            # Return the value for "b" when the above while loop is completed.

Number = 2                                              # Number whose square root is to be determined.
SQRT = str(sub_root(Number * 5, 5))                     # Set the value of "SQRT" to be the string extracted from the function "sub_root".
formatted_SQRT = f'{SQRT[0]}.{SQRT[1:101]}'             # Set the value of "format_SQRT to be the value of "SQRT" with a decimal place after the 1st digit in the string.
print("Square root 2 to 100 decimal places:")           # Print out the text.
print(formatted_SQRT)                                   # Print the value of "format_SQRT" to 100 decimal places.

Square root 2 to 100 decimal places:
1.4142135623730950488016887242096980785696718753769480731766797379907324784621070388503875343276415727


Actual square root of 2:

1.4142135623730950488016887242096980785696718753769480731766797379907324784621070388503875343276415727

https://apod.nasa.gov/htmltest/gifcity/sqrt2.1mil
The Square Root of Two to 1 Million Digits
Robert Nemiroff and Jerry Bonnell 


In [4]:
def sum_digits(digit):
    return sum(int(x) for x in digit if x.isdigit())

print(sum_digits(SQRT[:100]))

# https://stackoverflow.com/a/14550071/11485706
# Óscar López
# Jan 27 '13 at 17:32

475


https://projecteuler.net/problem=80

Square root digital expansion
Show HTML problem content 
Problem 80

It is well known that if the square root of a natural number is not an integer, then it is irrational. The decimal expansion of such square roots is infinite without any repeating pattern at all.

The square root of two is 1.41421356237309504880..., and the digital sum of the first one hundred decimal digits is 475.

For the first one hundred natural numbers, find the total of the digital sums of the first one hundred decimal digits for all the irrational square roots.

1.14 Decimals, Floats, and Floating Point Arithmetic. http://anh.cs.luc.edu/python/hands-on/3.1/handsonHtml/float.html
Hands-on-Python Tutorial
© Copyright 2019, Dr. Andrew N. Harrington. Last updated on Jan 05, 2020. Created using Sphinx 1.3.1+.
October 3rd, 2020

15. Floating Point Arithmetic: Issues and Limitations
The Python Tutorial
© Copyright 2001-2020, Python Software Foundation.
The Python Software Foundation is a non-profit corporation. Please donate.

Oct 08, 2020

In [5]:
Period of the Continued Fraction of √n
Marius Beceanu
February 5, 2003
https://web.archive.org/web/20151221205104/http://web.math.princeton.edu/mathlab/jr02fall/Periodicity/mariusjp.pdf

SyntaxError: invalid syntax (<ipython-input-5-479b27e948fe>, line 1)